design/loadbalancer-hash-policy-design.md

```diff
@@ -0,0 +1,185 @@
+
+```
+type LoadBalancerPolicy struct {
+    // Now also supports `RequestHash`
+	Strategy string `json:"strategy,omitempty"`
+
+	RequestHashPolicies []RequestHashPolicy `json:"requestHashPolicies,omitempty"`
+}
+```
+
+This field will be a list of `RequestHashPolicy` objects with each element holding configuration for an individual hash policy.
+Each list element will allow users configuring a `Route` on a `HTTPProxy` to specify a separate request attribute for Envoy to hash in order to make an upstream load balancing decision.
+`RequestHashPolicy` contains a field `RequestAttribute` specifying which type of request attribute it is targeting.
+If `RequestAttribute` is empty or an unknown value, this hash policy entry will be ignored.
+`HashOptions` is a generic `map[string]string` field that may contain options specific to the requested attribute.
+This may include what individual element of a set of properties on the request to use as an identifier to hash (for example the header or cookie name to hash the value of) and/or additional parameters to pass to Envoy to calculate a hash with (for example the TTL on a generated cookie).
+Depending on the `RequestAttribute`, some fields of this generic map may be required, an if missing, the hash policy excluded.
+For example, to implement header hashing, the `headerName` field would be required.
+The `Terminal` field denotes if the attribute specified (e.g. header name) is found, Envoy should stop processing any subsequent hash policies in the list it is provided.
+This is a performance optimization and can provide a speedup in hash calculation if set and the attribute to hash on is found.
+
+```
+type RequestHashPolicy struct {
+	RequestAttribute string            `json:"requestAttribute,omitempty"`
+	HashOptions      map[string]string `json:"hashOptions,omitempty"`
+	Terminal         bool              `json:"terminal,omitempty"`
+}
+```
+
+### HTTPProxy YAML Example
+
+An example of how this feature would be used to hash on a specific header value follows below:
+
+```
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+   name: example
+spec:
+  virtualhost:
+    fqdn: example.projectcontour.io
+  routes:
+  - services:
+    - name: example-app
+      port: 80
+    loadBalancerPolicy:
+      strategy: RequestHash
+      requestHashPolicies:
+      - requestAttribute: Header
+        hashOptions:
+          headerName: X-Some-Header
+        terminal: true
+      - requestAttribute: Header
+        hashOptions:
+          headerName: User-Agent
+```
+
+The value of the `X-Some-Header` header will be hashed by Envoy if present and used to make a load balancing decision.
+If the header is present, Envoy will not move on to attempting to hash the `User-Agent` header value.
+Consistent values in this header would lead to Envoy routing to the same backend service instance.
+
```

Comment:
also, I guess there's no way to specify a set of values to match on right, since this is on envoy side.  As in 'abc' and 'def' for a specified header both should hash to the same pod.

Reply:
Based on my understanding of the `hash_policy` docs, that's right.

Reply:
There is this option: https://www.envoyproxy.io/docs/envoy/v1.17.0/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-hashpolicy-header-regex-rewrite

But no one has asked for that yet, though I think that would do what youre asking: you would write a regex that matches two header values and the replacement value would be the same for both, would hash to the same thing etc.

Reply:
Thanks.  yea I was thinking sometimes you only have a referential identity, not a canonical one, as in I hold 'abc' and 'edf' to be equivalent and that's all that matters.  I guess you can always engineer a regex to encapsulate those two things